components/proto-derive/src/lib.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This crate provides macros for deriving some useful methods and traits for the exonum services.
+
+#![recursion_limit = "128"]
+#![deny(unsafe_code, bare_trait_objects)]
+#![warn(missing_docs, missing_debug_implementations)]
+
+extern crate proc_macro;
+
+mod pb_convert;
+
+use proc_macro::TokenStream;
+use syn::{Attribute, NestedMeta};
+
+/// ProtobufConvert derive macro.
+///
+/// Attributes:
+///
+/// ## Required
+///
+/// * `#[protobuf_convert(source = "path")]`
+///
+/// ```ignore
+/// #[derive(Clone, Debug, BinaryValue, ObjectHash, ProtobufConvert)]
+/// #[protobuf_convert(source = "proto::Wallet")]
+/// pub struct Wallet {
+///     /// `PublicKey` of the wallet.
+///     pub pub_key: PublicKey,
+///     /// Current balance of the wallet.
+///     pub balance: u64,
+/// }
+///
+/// let wallet = Wallet::new();
+/// let serialized_wallet = wallet.to_pb();
+///
+/// let deserialized_wallet = ProtobufConvert::from_pb(serialized_wallet).unwrap();
+/// assert_eq!(wallet, deserialized_wallet);
+/// ```
+///
+/// Corresponding proto file:
+/// ```text
+/// message Wallet {
+///  // Public key of the wallet owner.
+///  exonum.crypto.PublicKey pub_key = 1;
+///  // Current balance.
+///  uint64 balance = 2;
+/// }
+/// ```
+///
+/// This macro can also be applied to enums. In proto files enums are represented
+/// by `oneof` field. You can specify `oneof` field name, default is "message".
```

Comment:
Do I understand correctly that:

- The protobuf message must contain only this `oneof` field group (additional fields are not allowed)?
- The Rust enum should have variants with a single unnamed field? (Or are zero-field variants allowed too?)

It could be worth specifying these details.